### PR TITLE
Modular backend - inpaint

### PR DESCRIPTION
**Pull request by StAlKeR7779**
## Summary

Code for inpainting and inpaint models handling from https://github.com/invoke-ai/InvokeAI/pull/6577.
Separated in 2 extensions as discussed briefly before, so wait for discussion about such implementation.

## Related Issues / Discussions

#6606
https://invokeai.notion.site/Modular-Stable-Diffusion-Backend-Design-Document-e8952daab5d5472faecdc4a72d377b0d

## QA Instructions

Run with and without set `USE_MODULAR_DENOISE` environment.
Try and compare outputs between backends in cases:
- Normal generation on inpaint model
- Inpainting on inpaint model
- Inpainting on normal model

## Merge Plan

Nope.
If you think that there should be some kind of tests - feel free to add.

## Checklist

- [x] _The PR has a short but descriptive title, suitable for a changelog_
- [ ] _Tests added / updated (if applicable)_
- [ ] _Documentation added / updated (if applicable)_

### CODE REVIEW

File: invokeai/backend/stable_diffusion/extensions/inpaint.py
Comment (RyanJDick) [inline invokeai/backend/stable_diffusion/extensions/inpaint.py:16]:
```suggestion
class InpaintExt(ExtensionBase):
    """An extension for inpainting with non-inpainting models. See `InpaintModelExt` for inpainting with inpainting
    models.
    """
```

File: invokeai/backend/stable_diffusion/extensions/inpaint_model.py
Comment (RyanJDick) [inline invokeai/backend/stable_diffusion/extensions/inpaint_model.py:15]:
```suggestion
class InpaintModelExt(ExtensionBase):
    """An extension for inpainting with inpainting models. See `InpaintExt` for inpainting with non-inpainting
    models.
    """
```

File: invokeai/backend/stable_diffusion/extensions/inpaint.py
Comment (RyanJDick) [inline invokeai/backend/stable_diffusion/extensions/inpaint.py:55]: nit: raise `ValueError` to be slightly more specific.
```suggestion
            raise ValueError("InpaintExt should be used only on normal models!")
```

File: invokeai/backend/stable_diffusion/extensions/inpaint.py
Comment (RyanJDick) [inline invokeai/backend/stable_diffusion/extensions/inpaint.py:29]: Add a docstring explaining this check (i.e. that an inpainting model would have 9 in_channels).

File: invokeai/backend/stable_diffusion/extensions/inpaint.py
Comment (RyanJDick) [inline invokeai/backend/stable_diffusion/extensions/inpaint.py:66]: Add a comment explaining why we might have to re-generate the noise here. I think there is already one in the old code that can be ported over.

File: invokeai/backend/stable_diffusion/extensions/inpaint.py
Comment (RyanJDick) [inline invokeai/backend/stable_diffusion/extensions/inpaint.py:25]: Add documentation for the `self._noise` attribute to explain why it can't just be accessed from `ctx.inputs.noise`. (It's a bit of a hacky solution, so it's important that it is well-documented.)

File: invokeai/backend/stable_diffusion/extensions/inpaint.py
Comment (RyanJDick) [inline invokeai/backend/stable_diffusion/extensions/inpaint.py:21]: Add a docstring explaining these params. Here's a draft:
```suggestion
    def __init__(
        self,
        mask: torch.Tensor,
        is_gradient_mask: bool,
    ):
        """Initialize InpaintExt.
        Args:
            mask (torch.Tensor): The inpainting mask. Shape: (batch_size, 1, latent_height, latent_width). Values are
                expected to be in the range [0, 1]. A value of 0 means that the corresponding 'pixel' should not be
                inpainted.
            is_gradient_mask (bool): If True, mask is interpreted as a gradient mask meaning that the mask values range
                from 0 to 1. If False, mask is interpreted as binary mask meaning that the mask values are either 0 or
                1.
        """
```

I wrote this based on what I _thought_ these fields represented but based on testing this is wrong. The mask contains values in the range [0, 2]. Is this a bug in the code that prepares the mask? Or did I misunderstand the intended mask representation? We should get to the bottom of this.

File: invokeai/backend/stable_diffusion/extensions/inpaint.py
Comment (RyanJDick) [inline invokeai/backend/stable_diffusion/extensions/inpaint.py:45]: typo: threshhold -> threshold

File: invokeai/backend/stable_diffusion/extensions/inpaint.py
Comment (RyanJDick) [inline invokeai/backend/stable_diffusion/extensions/inpaint.py:46]: I'm not sure what this comment means, but we should figure it out (if there's something to figure out), and remove it.

File: invokeai/backend/stable_diffusion/extensions/inpaint.py
Comment (RyanJDick) [inline invokeai/backend/stable_diffusion/extensions/inpaint.py:69]: For this and POST_STEP, is there a specific interaction with another extension that you are trying to impose ordering on?

If there's no specific interaction that we are trying to control (yet), then I'd tend to just leave it as the default.

File: invokeai/backend/stable_diffusion/extensions/inpaint_model.py
Comment (RyanJDick) [inline invokeai/backend/stable_diffusion/extensions/inpaint_model.py:21]: Same comment as for `InpaintExt`, please add a docstring explaining these params since their meanings are non-obvious.

File: invokeai/backend/stable_diffusion/extensions/inpaint_model.py
Comment (RyanJDick) [inline invokeai/backend/stable_diffusion/extensions/inpaint_model.py:37]: nit: Raise ValueError to be slightly more specific.
```suggestion
            raise ValueError("InpaintModelExt should be used only on inpaint models!")
```

**Review by RyanJDick — CHANGES_REQUESTED** [review]
I tried to test, but ran into a device mismatch error. I expect that it will be easy to reproduce, but if not let me know and I can provide more details.

Also, in addition to the test cases that you have already listed, we should make sure and test both gradient-mask and non-gradient-mask inpainting.

File: invokeai/backend/stable_diffusion/extensions/inpaint.py
Comment (StAlKeR7779) [reply invokeai/backend/stable_diffusion/extensions/inpaint.py:21]: It's because gradient mask outputs mask in range [-1,1] I'll include fix of this in this PR.

File: invokeai/backend/stable_diffusion/extensions/inpaint.py
Comment (StAlKeR7779) [reply invokeai/backend/stable_diffusion/extensions/inpaint.py:69]: I not sure, as here we simply mean - run earlier than ANY default extension.

**Review by lstein — COMMENTED** [review]
> ```
> // invokeai/app/invocations/denoise_latents.py 
>  def invoke(self, context: InvocationContext) -> LatentsOutput:
>         if os.environ.get("USE_MODULAR_DENOISE", False):
>             return self._new_invoke(context)
>         else:
>             return self._old_invoke(context)
> ```

During the transition period, why not make `USE_MODULAR_DENOISE` into an app configuration variable? This will then automatically set a config parameter based on an environment variable named `INVOKEAI_USE_MODULAR_DENOISE` and be consistent with how environment variables are handled elsewhere.

**Comment by lstein** [comment]
Ran into an apparent bug. When inpainting with `stable-diffusion-xl-1.0-inpainting-0.1`:
```
$ USE_MODULAR_DENOISE="1" invokeai-web
  File "/home/lstein/Projects/InvokeAI/invokeai/app/services/session_processor/session_processor_default.py", line 129, in run_node
    output = invocation.invoke_internal(context=context, services=self._services)
  File "/home/lstein/Projects/InvokeAI/invokeai/app/invocations/baseinvocation.py", line 289, in invoke_internal
    output = self.invoke(context)
  File "/home/lstein/Projects/InvokeAI/invokeai/app/invocations/denoise_latents.py", line 722, in invoke
    return self._new_invoke(context)
  File "/home/lstein/invokeai-main/.venv/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/usr/lib/python3.10/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/home/lstein/Projects/InvokeAI/invokeai/app/invocations/denoise_latents.py", line 818, in _new_invoke
    result_latents = sd_backend.latents_from_embeddings(denoise_ctx, ext_manager)
  File "/home/lstein/Projects/InvokeAI/invokeai/backend/stable_diffusion/diffusion_backend.py", line 45, in latents_from_embeddings
    ext_manager.run_callback(ExtensionCallbackType.PRE_DENOISE_LOOP, ctx)
  File "/home/lstein/Projects/InvokeAI/invokeai/backend/stable_diffusion/extensions_manager.py", line 52, in run_callback
    cb.function(ctx)
  File "/home/lstein/Projects/InvokeAI/invokeai/backend/stable_diffusion/extensions/inpaint.py", line 77, in init_tensors
    raise ValueError("InpaintExt should be used only on normal models!")
ValueError: InpaintExt should be used only on normal models!
```
No such issue when `USE_MODULAR_DENOISE` is false. I checked the `config.json` for the unet, and it is 9 channels as expected for an inpainting model.

**Comment by StAlKeR7779** [comment]
> Ran into an apparent bug. When inpainting with `stable-diffusion-xl-1.0-inpainting-0.1`:
> 
> ```
> $ USE_MODULAR_DENOISE="1" invokeai-web
> ... multiple log lines omitted...
>   File "/home/lstein/Projects/InvokeAI/invokeai/app/services/session_processor/session_processor_default.py", line 129, in run_node
>     output = invocation.invoke_internal(context=context, services=self._services)
>   File "/home/lstein/Projects/InvokeAI/invokeai/app/invocations/baseinvocation.py", line 289, in invoke_internal
>     output = self.invoke(context)
>   File "/home/lstein/Projects/InvokeAI/invokeai/app/invocations/denoise_latents.py", line 722, in invoke
>     return self._new_invoke(context)
>   File "/home/lstein/invokeai-main/.venv/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
>     return func(*args, **kwargs)
>   File "/usr/lib/python3.10/contextlib.py", line 79, in inner
>     return func(*args, **kwds)
>   File "/home/lstein/Projects/InvokeAI/invokeai/app/invocations/denoise_latents.py", line 818, in _new_invoke
>     result_latents = sd_backend.latents_from_embeddings(denoise_ctx, ext_manager)
>   File "/home/lstein/Projects/InvokeAI/invokeai/backend/stable_diffusion/diffusion_backend.py", line 45, in latents_from_embeddings
>     ext_manager.run_callback(ExtensionCallbackType.PRE_DENOISE_LOOP, ctx)
>   File "/home/lstein/Projects/InvokeAI/invokeai/backend/stable_diffusion/extensions_manager.py", line 52, in run_callback
>     cb.function(ctx)
>   File "/home/lstein/Projects/InvokeAI/invokeai/backend/stable_diffusion/extensions/inpaint.py", line 77, in init_tensors
>     raise ValueError("InpaintExt should be used only on normal models!")
> ValueError: InpaintExt should be used only on normal models!
> ```
> 
> No such issue when `USE_MODULAR_DENOISE` is false. I checked the `config.json` for the unet, and it is 9 channels as expected for an inpainting model.

Could you debug if `unet_config.variant` correct on this model?
Because currently in code `InpaintExt`/`InpaintModelExt` selects based on `unet_config.variant`, while internally code checks for `in_channels`(and error from this check you see).

**Comment by lstein** [comment]
> Could you debug if `unet_config.variant` correct on this model?
> Because currently in code `InpaintExt`/`InpaintModelExt` selects based on `unet_config.variant`, while internally code checks for `in_channels`(and error from this check you see).

You called it correctly. The model was configured as "normal". Changing its variant to "inpaint" fixed the issue.

Now I have to figure out why the variant was set to normal - maybe an issue with the prober.

**Review by RyanJDick — COMMENTED** [review]
It might make more sense to split the CreateGradientMaskInvocation bug fix out into its own PR. What do you think? This would simplify the testing scope a bit. And will certainly be nice if a rollback is necessary. I'll leave it up to you. If we keep the bug fix in this PR, can you update the PR description to reflect that it also fixes a bug in CreateGradientMaskInvocation - which will have a small impact on behavior.

**Comment by RyanJDick** [comment]
We need to test all of the following cases for regression against `main` before merging:
- [ ] _old_invoke, gradient mask, inpaint model
- [ ] _old_invoke, gradient mask, non-inpaint model
- [ ] _old_invoke, binary mask, inpaint mode
- [ ] _old_invoke, binary mask, non-inpaint mode
- [ ] _old_invoke, normal generation with an inpainting model
- [ ] _new_invoke, gradient mask, inpaint model
- [ ] _new_invoke, gradient mask, non-inpaint model
- [ ] _new_invoke, binary mask, inpaint mode
- [ ] _new_invoke, binary mask, non-inpaint mode
- [ ] _new_invoke, normal generation with an inpainting model

**Comment by RyanJDick** [comment]
> During the transition period, why not make `USE_MODULAR_DENOISE` into an app configuration variable? This will then automatically set a config parameter based on an environment variable named `INVOKEAI_USE_MODULAR_DENOISE` and be consistent with how environment variables are handled elsewhere.

I don't feel strongly either way. But, I didn't suggest it earlier since it's intended to be short-lived and for developers only, do I didn't want to have to worry about handling config migrations.

File: invokeai/backend/stable_diffusion/extensions/inpaint.py
Comment (StAlKeR7779) [reply invokeai/backend/stable_diffusion/extensions/inpaint.py:94]: At least in POST_STEP, current preview depends on order.

File: invokeai/app/invocations/create_gradient_mask.py
Comment (StAlKeR7779) [reply invokeai/app/invocations/create_gradient_mask.py:96]: Originally @dunkeroni thought that negative values will be ignored(treated same as 0), at least it how I understand him.
From this point of view... I see it at least not as major version update(like `"1.2.0"`).
Also I can add that else branch outputs [0, 1], only positive if branch in main outputs [-1, 1], so I see it more as bug.

File: invokeai/app/invocations/denoise_latents.py
Comment (StAlKeR7779) [reply invokeai/app/invocations/denoise_latents.py:834]: I thought about some comments, but this code intended to be removed later, so I not added.
Also I think comment should more like: `We invert mask here for compatibility with old backend implementation, which uses inversed mask.`

File: invokeai/app/invocations/create_gradient_mask.py
Comment (StAlKeR7779) [reply invokeai/app/invocations/create_gradient_mask.py:96]: And yeah, I'll move it to separate PR.

**Comment by StAlKeR7779** [comment]
> We need to test all of the following cases for regression against `main` before merging:
> 
> * [ ]  _old_invoke, gradient mask, inpaint model
> * [ ]  _old_invoke, gradient mask, non-inpaint model
> * [ ]  _old_invoke, binary mask, inpaint mode
> * [ ]  _old_invoke, binary mask, non-inpaint mode
> * [ ]  _old_invoke, normal generation with an inpainting model
> * [ ]  _new_invoke, gradient mask, inpaint model
> * [ ]  _new_invoke, gradient mask, non-inpaint model
> * [ ]  _new_invoke, binary mask, inpaint mode
> * [ ]  _new_invoke, binary mask, non-inpaint mode
> * [ ]  _new_invoke, normal generation with an inpainting model

Runned all this cases - all looks ok

File: invokeai/backend/stable_diffusion/extensions/inpaint_model.py
Comment (dunkeroni) [inline invokeai/backend/stable_diffusion/extensions/inpaint_model.py:88]: Mask>0 is slightly larger than the VAE-encoded mask area, which has its boundary at 0.5 instead (from the create denoise mask nodes). That can cause fringe artifacts on the mask edge. I think it would be better to ignore is_gradient_mask entirely when dealing with inpaint models and always treat them with a .lerp instead of choosing.

Comment (StAlKeR7779) [reply invokeai/backend/stable_diffusion/extensions/inpaint_model.py:88]: After some discussion in discord decided to do it later in separated PR.
To not have multiple changes-tests in this PR.

File: invokeai/backend/stable_diffusion/extensions/inpaint.py
Comment (RyanJDick) [reply invokeai/backend/stable_diffusion/extensions/inpaint.py:98]: It seems a little less clear-cut than the preview case, but I don't feel strongly. If you keep it, can you just add your explanation as a comment above that line and remove the `TODO`?

Same with all the other TODOs around order values - best to just add a comment explaining why the current  decision was made and then they can be updated easily in the future.

**Review by RyanJDick — APPROVED** [review]

**Review by hipsterusername — APPROVED** [review]

**Pull request merged**
